Mathlib/Data/Fintype/EquivFin.lean

```diff
@@ -0,0 +1,684 @@
+
+See `Fintype.equivFinOfCardEq` for the noncomputable definition,
+and `Fintype.truncEquivFin` and `Fintype.equivFin` for the bijection `α ≃ Fin (card α)`.
+-/
+def truncEquivFinOfCardEq [DecidableEq α] {n : ℕ} (h : Fintype.card α = n) : Trunc (α ≃ Fin n) :=
+  (truncEquivFin α).map fun e => e.trans (finCongr h)
+
+/-- If the cardinality of `α` is `n`, there is noncomputably a bijection between `α` and `Fin n`.
+
+See `Fintype.truncEquivFinOfCardEq` for the computable definition,
+and `Fintype.truncEquivFin` and `Fintype.equivFin` for the bijection `α ≃ Fin (card α)`.
+-/
+noncomputable def equivFinOfCardEq {n : ℕ} (h : Fintype.card α = n) : α ≃ Fin n :=
+  letI := Classical.decEq α
+  (truncEquivFinOfCardEq h).out
+
+/-- Two `Fintype`s with the same cardinality are (computably) in bijection.
+
+See `Fintype.equivOfCardEq` for the noncomputable version,
+and `Fintype.truncEquivFinOfCardEq` and `Fintype.equivFinOfCardEq` for
+the specialization to `Fin`.
+-/
+def truncEquivOfCardEq [DecidableEq α] [DecidableEq β] (h : card α = card β) : Trunc (α ≃ β) :=
+  (truncEquivFinOfCardEq h).bind fun e => (truncEquivFin β).map fun e' => e.trans e'.symm
+
+/-- Two `Fintype`s with the same cardinality are (noncomputably) in bijection.
+
+See `Fintype.truncEquivOfCardEq` for the computable version,
+and `Fintype.truncEquivFinOfCardEq` and `Fintype.equivFinOfCardEq` for
+the specialization to `Fin`.
+-/
+noncomputable def equivOfCardEq (h : card α = card β) : α ≃ β := by
+  letI := Classical.decEq α
+  letI := Classical.decEq β
+  exact (truncEquivOfCardEq h).out
+
+end
+
+theorem card_eq {α β} [_F : Fintype α] [_G : Fintype β] : card α = card β ↔ Nonempty (α ≃ β) :=
+  ⟨fun h =>
+    haveI := Classical.propDecidable
+    (truncEquivOfCardEq h).nonempty,
+    fun ⟨f⟩ => card_congr f⟩
+
+end Fintype
+
+theorem Fintype.card_fin_lt_of_le {m n : ℕ} (h : m ≤ n) :
+    Fintype.card {i : Fin n // i < m} = m := by
+  conv_rhs => rw [← Fintype.card_fin m]
+  apply Fintype.card_congr
+  exact { toFun := fun ⟨⟨i, _⟩, hi⟩ ↦ ⟨i, hi⟩
+          invFun := fun ⟨i, hi⟩ ↦ ⟨⟨i, lt_of_lt_of_le hi h⟩, hi⟩
+          left_inv := fun i ↦ rfl
+          right_inv := fun i ↦ rfl }
+
+theorem Finset.card_fin (n : ℕ) : #(univ : Finset (Fin n)) = n := by simp
+
+/-- `Fin` as a map from `ℕ` to `Type` is injective. Note that since this is a statement about
+equality of types, using it should be avoided if possible. -/
+theorem fin_injective : Function.Injective Fin := fun m n h =>
+  (Fintype.card_fin m).symm.trans <| (Fintype.card_congr <| Equiv.cast h).trans (Fintype.card_fin n)
+
+theorem Fin.val_eq_val_of_heq {k l : ℕ} {i : Fin k} {j : Fin l} (h : HEq i j) :
+    (i : ℕ) = (j : ℕ) :=
+  (Fin.heq_ext_iff (fin_injective (type_eq_of_heq h))).1 h
+
+/-- A reversed version of `Fin.cast_eq_cast` that is easier to rewrite with. -/
+theorem Fin.cast_eq_cast' {n m : ℕ} (h : Fin n = Fin m) :
+    _root_.cast h = Fin.cast (fin_injective h) := by
+  cases fin_injective h
+  rfl
```

Comment:
All of this can be moved earlier, I believe

Reply:
You are completely correct, these now live together with `Fintype.card_fin` into their own section in `Fintype.Card`.